analytics/src/test/java/com/segment/analytics/internal/AnalyticsClientTest.java

```diff
@@ -76,13 +79,48 @@ AnalyticsClient newClient() {
   }
 
   @Test
-  public void enqueueAddsToQueue(MessageBuilderTest builder) throws InterruptedException {
+  public void enqueueAddsToQueue(MessageBuilderTest builder) {
     AnalyticsClient client = newClient();
 
     Message message = builder.get().userId("prateek").build();
     client.enqueue(message);
 
-    verify(messageQueue).put(message);
+    verify(messageQueue).offer(message);
+  }
+
+  @Test
+  public void enqueueDropsMessageWhenQueueFull(MessageBuilderTest builder) {
+    BlockingQueue messageQueue = new LinkedBlockingQueue<Message>(10);
+    // simulate indefinite blocking on submission to network executor
+    doAnswer(new Answer() {
+      @Override
+      public Object answer(InvocationOnMock invocation) throws Throwable {
+        Thread.sleep(3600 * 1000);
+        return null;
+      }
+    }).when(networkExecutor).submit(any(Runnable.class));
+
+    AnalyticsClient analyticsClient = new AnalyticsClient(
+      messageQueue,
+      segmentService,
+      10,
+      TimeUnit.HOURS.toMillis(1),
+      log,
+      threadFactory,
+      networkExecutor,
+      Collections.singletonList(callback));
+
+    for (int i = 0; i < 25; i++) {
+      Message message = builder.get().userId("prateek").build();
+      analyticsClient.enqueue(message);
+    }
+
+    // 10 messages are removed by looperExecutor,
+    // 10 messages are retained,
+    // 5 are dropped.
+    assertThat(messageQueue).hasSize(10);
+    verify(log, times(5))
+      .print(Log.Level.ERROR, "Failed to enqueue message as queue is already full.");
```

Comment:
I'm wondering if checking if the log was called is a proper test assertion here.